<center>
<h1> TP-Projet d'optimisation numérique </h1>
<h1> Année 2020-2021 - 2e année département Sciences du Numérique </h1>
<h1> Noms: DAI </h1>
<h1> Prénoms: Guohao </h1>    
</center>

# Algorithme de Newton
## Implémentation 
 
1. Coder l’algorithme de Newton local tel que décrit dans la section *Algorithme de Newton* (fichier `Algorithme_De_Newton.jl`)

2. Tester l’algorithme sur les fonctions $f_{1}$ , $f_{2}$ avec les points initiaux $x_{011}$ , $x_{012}$ (pour $f_{1}$ ) et $x_{021}$ , $x_{022}$ , $x_{023}$ (pour $f_{2}$ ) donnés en Annexe A.

In [44]:
using Pkg
Pkg.status()

      Status `~/.julia/environments/v1.7/Project.toml`
  [bd48cda9] GraphRecipes v0.5.8
  [7073ff75] IJulia v1.23.2
  [91a5bcdd] Plots v1.24.1
  [2913bbd2] StatsBase v0.33.13
  [a6016688] TestOptinum v0.1.0 `https://github.com/mathn7/TestOptinum.git#master`
  [37e2e46d] LinearAlgebra
  [d6f4376e] Markdown


In [45]:
#using Pkg; Pkg.add("LinearAlgebra"); Pkg.add("Markdown")
# using Documenter
using LinearAlgebra
using Markdown                             # Pour que les docstrings en début des fonctions ne posent
                                           # pas de soucis. Ces docstrings sont utiles pour générer 
                                           # la documentation sous GitHub
include("Algorithme_De_Newton.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,nom_fct,point_init,xmin,fxmin,flag,sol_exacte,nbiters)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à ",nom_fct, " au point initial ", point_init, ":\n",bold=true,color=:blue)
	println("  * xsol = ",xmin)
	println("  * f(xsol) = ",fxmin)
	println("  * nb_iters = ",nbiters)
	println("  * flag = ",flag)
	println("  * sol_exacte : ", sol_exacte)
end

# Fonction f0
# -----------
f0(x) =  sin(x)
# la gradient de la fonction f0
grad_f0(x) = cos(x)
# la hessienne de la fonction f0
hess_f0(x) = -sin(x)
sol_exacte = -pi/2
options = []

println("* Newton: Test of f0 *")
x0 = sol_exacte
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = -pi/2+0.5
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = pi/2
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)


* Newton: Test of f0 *
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.5707963267948966:
  * xsol = -1.5707963267948966
  * f(xsol) = -1.0
  * nb_iters = 0
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.0707963267948966:
  * xsol = -1.5707963267949088
  * f(xsol) = -1.0
  * nb_iters = 3
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial 1.5707963267948966:
  * xsol = 1.5707963267948966
  * f(xsol) = 1.0
  * nb_iters = 0
  * flag = 0
  * sol_exacte : -1.5707963267948966


In [ ]:
##### TODO : Fonction f1 and f2 according to Annexes

# Fonction f1
# -----------
# function f1(x1, x2, x3) -> 2*(x1+x2+x3-3)^2 + (x1-x2)^2 + (x2-x3)^2
f1(x) = 2 * (x[1] + x[2] + x[3] - 3) ^ 2 + (x[1] - x[2]) ^ 2 + ( x[2] - x[3]) ^ 2
# la gradient de la fonction f1 ->  grad(2*(x1+x2+x3-3)^2 + (x1-x2)^2 + (x2-x3)^2)
grad_f1(x) = [4*(x[1]+x[2]+x[3]-3) + 2*(x[1]-x[2]); 4*(x[1]+x[2]+x[3]-3) - 2*(x[1]-x[2])+2*(x[2]-x[3]); 4*(x[1]+x[2]+x[3]-3)-2*(x[2]-x[3])]
# la hessienne de la fonction f1
hess_f1(x) = [6 2 4; 2 8 2; 4 2 6]
sol_exacte = [1;1;1]
options = []

println("\n\n* Newton: Test of f1 * ")
x1_1 = [1; 0; 0]
# println("test1")
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f1,grad_f1,hess_f1,x1_1,options)
my_afficher_resultats("Newton","f1",x1_1,xmin,f_min,flag,sol_exacte,nb_iters)
# println("tets2")
x1_2 = [10; 3; -2.2]
# println("test3")
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f1,grad_f1,hess_f1,x1_2,options)
my_afficher_resultats("Newton","f1",x1_2,xmin,f_min,flag,sol_exacte,nb_iters)


In [ ]:
# Fonction f2
# -----------
# function f2(x1, x2) -> 100*(x2-(x1^2))^2 + (1-x1)^2
f2(x) = 100 * (x[2] - (x[1] ^ 2)) ^ 2 + (1 - x[1]) ^ 2
# la gradient de la fonction f2
grad_f2(x) = [-400*x[1]*(x[2]-x[1]^2)-2*(1-x[1]) ; 200*(x[2]-x[1]^2)]
# la hessienne de la fonction f2
hess_f2(x) = [1200*x[1]^2-400*x[2]+2 (-400*x[1]); -400*x[1] 200]
sol_exacte = [1;1]
options = []
println(hess_f2([1;1]))
println("\n\n* Newton: Test of f2 *")
x2_1 = [-1.2; 1]
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f2,grad_f2,hess_f2,x2_1,options)
my_afficher_resultats("Newton","f2",x2_1,xmin,f_min,flag,sol_exacte,nb_iters)

x2_2 = [10; 0]
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f2,grad_f2,hess_f2,x2_2,options)
my_afficher_resultats("Newton","f2",x2_2,xmin,f_min,flag,sol_exacte,nb_iters)

x2_3 = [0; (1/200)+1/(10^12)]
println("# checkpoint 1")
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f2,grad_f2,hess_f2,x2_3,options)
println("# checkpoint 2")
my_afficher_resultats("Newton","f2",x2_3,xmin,f_min,flag,sol_exacte,nb_iters)

## Interprétation 

Justifier

1. les résultats obtenus pour l'exemple $f_0$ ci-dessus;

2. que l’algorithme implémenté converge en une itération pour $f_{1}$;

3. que l’algorithme puisse ne pas converger pour $f_{2}$ avec certains points initiaux.

## Réponses:

1. In the first two tests of function $f_0 = sin (x)$, when the initial value is $-1.5707963267948966$, and the initial value is $-1.0707963267948966$, there is a minimum value, and the result also converges to the minimum value. Their minimum result is the same, which is $-1$. In the last test, when the initial value is $1.5707963267948966$, we get the maximum value of the sine function, which is +1.

2. Because it is equal to its second-order Taylor expansion, the minimum value can be obtained in a single iteration.

3. As shown in the third test of $f_2 $, at some initial points of $f_2 $, Newton's algorithm may not converge. This is because approximating the extreme value of the function through the second-order Taylor expansion of $ f_2 $ will cause the function to lead to the minimum in the opposite direction. The resulting system may be irreversible.


# Régions de confiance avec pas de cauchy 

## Implémentation 

1. Coder l'algorithme du pas de Cauchy d’un sous-problème de
régions de confiance (fichier `Pas_De_Cauchy.jl`). Tester sur les quadratiques proposées en Annexe B.

2. Coder l'algorithme de régions de confiance (fichier `Regions_De_Confiance.jl`). Tester sur les problèmes de l’Annexe A.

In [ ]:
# Vos tests
##### TODO : Test of "Pas de Cauchy"
include("Pas_De_Cauchy.jl")

# On considère des fonctions quadratiques de la forme q(s) = s'*g + 0.5*s'*H*s

# Quadratique 1
g1 = [0; 0]
H1 = [7 0; 0 2]
delta1 = 1
solution1, state1 = Pas_De_Cauchy(g1,H1,delta1)
println("\n* Pas de Cauchy: Test of Quadratique 1 *")
println("-------------------------------------------------------------------------")
println("* The approximate solution of Quadratique 1 is ", solution1)
println("* The state of exit is ", state1)


# Quadratique 2
g2 = [6; 2]
H2 = [7 0; 0 2]
delta2 = 1
solution2, state2 = Pas_De_Cauchy(g2,H2,delta2)
println("\n\n* Pas de Cauchy: Test of Quadratique 2 *")
println("-------------------------------------------------------------------------")
println("* The approximate solution of Quadratique 2 is ", solution2)
println("* The state of exit is ", state2)


# Quadratique 3
g3 = [-2; 1]
H3 = [-2 0; 0 10]
delta3 = 1
solution3, state3 = Pas_De_Cauchy(g3,H3,delta3)
println("\n\n* Pas de Cauchy: Test of Quadratique 3 *")
println("-------------------------------------------------------------------------")
println("* The approximate solution of Quadratique 3 is ", solution3)
println("* The state of exit is ", state3,"\n\n")

## Interprétation 

1. Quelle relation lie la fonction test $f_1$ et son modèle de Taylor à l’ordre 2 ? Comparer alors les performances de Newton et RC-Pas de Cauchy sur cette fonction.

2. Le rayon initial de la région de confiance est un paramètre important dans l’analyse
de la performance de l’algorithme. Sur quel(s) autre(s) paramètre(s) peut-on jouer
pour essayer d’améliorer cette performance ? Étudier l’influence d’au moins deux de
ces paramètres.

## Réponses:
1. For $f_1$ :  
   $f_1$ is a quadratic function, so it is equal to its second-order Taylor expansion, so the Newton method converges in one iteration, and its accuracy reaches $e^{32}$.  
   In the other hand, RC-Pas de Cauchy method converges to a less precise result ($e^{16}$)in 49 iterations.  
   Therefore, Newton method is more efficient in $f_1$ because it obtains more accurate results at a lower cost.

2. We can add these parameters, $\Delta_{max}$, $\gamma_{min}$, $\gamma_{max}$, $\eta_{min}$ and $\eta_{max}$, to improve the efficiency of the algorithm: 
   - $\Delta_{max}$ : Maximum radius of confience area;
   - $\gamma_{min}$ and $\gamma_{max}$ : the factors for the expansion and contraction of the confidence area;
   - $\eta_{min}$ and $\eta_{max}$ : the criteria for the expansion and contraction of the confidence area.


   When $\gamma_{max}$ increases, it will gradually approach $1$, and $\rho$ will be less than $\gamma_{max}$. Then $\delta$ will not change. When $\gamma_{max}$ decreases, $\gamma_{max}$ will be close to $\gamma_{min}$.

   When $\eta_{min}$ decreases, it will approach to $0$, and $\gamma_{min}$ will decrease, then $\delta$ approaches $0$.

# Régions de confiance avec gradient conjugué tronqué

## Implémentation 

1. Implémenter l’algorithme du Gradient Conjugué Tronqué, en se basant sur le cours (fichier `Gradient_Conjugue_Tronque.jl`).
On validera les résultats sur les fonctions de l’Annexe C.

2. Intégrer finalement l’algorithme du Gradient Conjugué Tronqué dans le code de
régions de confiance, et appliquer ce code pour résoudre les exemples proposés en
Annexe A.

In [ ]:
# Vos tests
include("Regions_De_Confiance.jl")
include("Pas_De_Cauchy.jl")
include("Gradient_Conjugue_Tronque.jl")

###### TODO: 
# 1. Implémenter l’algorithme du Gradient Conjugué Tronqué
printstyled("\nSolution 1 : Implémenter l’algorithme du Gradient Conjugué Tronqué\n", bold=true, color=:blue, underline=true)

# Quadratique 4
g4 = [0; 0]
H4 = [-2 0; 0 10]
options = []
solution4 = Gradient_Conjugue_Tronque(g4,H4,options)
println("\n* Gradient Conjugue Tronque: Test of Quadratique 4 *")
println("-------------------------------------------------------------------------")
println("* The approximate solution of Quadratique 4 is ", solution4)

# Quadratique 5
g5 = [2; 3]
H5 = [4 6; 6 5]
solution5 = Gradient_Conjugue_Tronque(g5,H5,options)
println("\n\n* Gradient Conjugue Tronque: Test of Quadratique 5 *")
println("-------------------------------------------------------------------------")
println("* The approximate solution of Quadratique 5 is ", solution5)

# Quadratique 6
g6 = [2; 0]
H6 = [4 0; 0 -15]
solution6 = Gradient_Conjugue_Tronque(g6,H6,options)
println("\n\n* Gradient Conjugue Tronque: Test of Quadratique 6 *")
println("-------------------------------------------------------------------------")
println("* The approximate solution of Quadratique 6 is ", solution6, "\n\n\n")




In [ ]:
###### TODO: 
# Vos tests
include("Regions_De_Confiance.jl")
include("Pas_De_Cauchy.jl")
include("Gradient_Conjugue_Tronque.jl")

# 2. L’algorithme du Gradient Conjugué Tronqué dans Regions De Confiance pour résoudre les exemples proposés en Annexe A.
printstyled("\nSolution 2 : L’algorithme du Gradient Conjugué Tronqué dans Regions De Confiance pour résoudre les exemples proposés en Annexe A.\n", bold=true, color=:blue, underline=true)

# Fonction f1
# -----------
sol_exacte = [1;1;1]

x1_1 = [1; 0; 0]

xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f1,grad_f1,hess_f1,x1_1,options)
my_afficher_resultats("gct","f1",x1_1,xmin,f_min,flag,sol_exacte,nb_iters)

x1_2 = [10; 3; -2.2]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f1,grad_f1,hess_f1,x1_2,options)
my_afficher_resultats("gct","f1",x1_2,xmin,f_min,flag,sol_exacte,nb_iters)

In [ ]:
sol_exacte = [1;1]
options = []

println("\n\n* Newton: Test of f2 *")
x2_1 = [-1.2; 1]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f2,grad_f2,hess_f2,x2_1,options)
my_afficher_resultats("gct","f2",x2_1,xmin,f_min,flag,sol_exacte,nb_iters)

x2_2 = [10; 0]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f2,grad_f2,hess_f2,x2_2,options)
my_afficher_resultats("gct","f2",x2_2,xmin,f_min,flag,sol_exacte,nb_iters)

x2_3 = [0; ((1/200)+(1/(10^12)))]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f2,grad_f2,hess_f2,x2_3,options)
my_afficher_resultats("gct","f2",x2_3,xmin,f_min,flag,sol_exacte,nb_iters)

## Interprétation  

1. Comparer la décroissance obtenue avec celle du pas de Cauchy, en retournant, dans
un premier temps le dernier itéré admissible à courbure positive (c’est à dire, que si
l’une ou l’autre des deux conditions (b) ou (d) sont rencontrées dans l’algorithme 3,
alors on ne calcule pas ``σ_{j}`` et on retourne le dernier itéré ``s_{j}`` directement).

2. Comparer la décroissance obtenue avec celle du pas de Cauchy, en imposant la sortie
dans l’algorithme 3 au bout d’une itération seulement. Que remarquez vous ?

3. Comparer la décroissance obtenue avec celle du pas de Cauchy dans le cas général.

4. Quels sont les avantages et inconvénients des deux approches ?

## Réponses:
Based on the above test results, we can get :
- **Pas de Cauchy** : In some simple problems, it can give a result with sufficient accuracy with a very small number of iterations; but in general, it is not very accurate and slower than "gct", even in a given maximum number of iterations. Accurate results
- **gct** : In most cases, the calculation steps and directions yield results faster with fewer iterations, and provide more accurate solutions than "Pas de Cauchy" steps


# Lagrangien augmenté

## Implémentation
1.Choisir des critères d’arrêt pour la convergence de l'algorithme.

2.Implémenter l'algorithme du lagrangien augmenté, en utilisant les différentes méthodes
qui ont été vues en première partie pour la résolution de la suite de problémes sans
contraintes (fichier `Lagrangien_Augmente.jl`)
 
3.Tester les différentes variantes sur les problèmes en Annexe D.
 

In [ ]:
# Vos tests
###### TODO: On s'intéresse à la valeur minimale de f1 sur un ensemble défini par une contrainte linéaire.
###### Annexe D. Problèmes avec contraintes

include("Lagrangien_Augmente.jl")

printstyled("\nAnnexe D. Problèmes avec contraintes : Retour sur f1 , réalisable\n", bold=true, color=:blue, underline=true)

# Retour sur f1
contrainte(x) = x[1]+x[3]-1
grad_contrainte(x) = [1; 0; 1]
hess_contrainte(x) = [0 0 0; 0 0 0; 0 0 0]

# réalisable
xc = [0; 1; 1]
sol_exacte = [1;1;1]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : newton","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("gct",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : gct","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("cauchy",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : cauchy","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)



In [ ]:
printstyled("\nAnnexe D. Problèmes avec contraintes : Retour sur f1 , non-réalisable\n", bold=true, color=:blue, underline=true)

# non réalisable
xc = [0.5; 1.25; 1]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : newton","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("gct",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : gct","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("cauchy",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : cauchy","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)


In [ ]:
printstyled("\nAnnexe D. Problèmes avec contraintes : Retour sur f2 , réalisable\n", bold=true, color=:blue, underline=true)
sol_exacte = [1;1]
# Retour sur f2
contrainte(x) =  (x[1]^2) + (x[2]^2) -1.5
grad_contrainte(x) = [2 * x[1]; 2 * x[2]]
hess_contrainte(x) = [2 0; 0 2]

#réalisable
xc = [sqrt(3)/2; sqrt(3)/2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : newton","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("gct",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : gct","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("cauchy",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : cauchy","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)



In [ ]:
printstyled("\nAnnexe D. Problèmes avec contraintes : Retour sur f2 , non-réalisable\n", bold=true, color=:blue, underline=true)

# non réalisable
xc = [1; 0]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton", f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente (newton)","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("gct",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : gct","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("cauchy",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente : cauchy","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)



In [ ]:
#=
* **options**      : (Array{Float,1})
1. **epsilon** 	   : utilisé dans les critères d'arrêt
2. **tol**         : la tolérance utilisée dans les critères d'arrêt
3. **itermax** 	   : nombre maximal d'itération dans la boucle principale
4. **lambda0**	   : la deuxième composante du point de départ du Lagrangien
5. **mu0,tho** 	   : valeurs initiales des variables de l'algorithme 

        epsilon = 1e-8      options[1]		tol_abs
		tol = 1e-5          options[2]		tol_nel
		itermax = 1000      options[3]
		lambda0 = 2         options[4]
		mu0 = 100           options[5]
		tho = 2             options[6]
=#

function afficher_comparison_resultats(algo,nom_fct,point_init,itermax,lambda0,mu0,tho,xmin,fxmin,flag,sol_exacte,nbiters)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à ",nom_fct, " au point initial ", point_init, " itermax = ",itermax,":\n",bold=true,color=:blue)
    printstyled("  * lambda0 = ",lambda0,"\n",color=:red)
	printstyled("  * mu0 = ",mu0,"\n",color=:red)
    printstyled("  * tho = ",tho,"\n",color=:red)
	println("  * xsol = ",xmin)
	println("  * f(xsol) = ",fxmin)
	println("  * nb_iters = ",nbiters)
	println("  * flag = ",flag)
end

xc = [sqrt(3)/2; sqrt(3)/2]
### DEFAULT
options = [1e-8 1e-5 1000 2 100 2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)

println()
options = [1e-8 1e-5 1000 0.001 100 2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)

println()
options = [1e-8 1e-5 1000 1000 100 2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)



In [ ]:
#### DEFAULT
println()
options = [1e-8 1e-5 1000 2 100 2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)

println()
options = [1e-8 1e-5 1000 2 0.1 2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)

println()
options = [1e-8 1e-5 1000 2 1 2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)

println()
options = [1e-8 1e-5 1000 2 10000 2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)

In [ ]:
##### DEFAULT
println()
options = [1e-8 1e-5 1000 2 100 2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)

println()
options = [1e-8 1e-5 1000 2 100 0.02]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)

println()
options = [1e-8 1e-5 1000 2 100 20000]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,options)
afficher_comparison_resultats("Lagrangien_Augmente : newton","f2",xc,options[3],options[4],options[5],options[6],xmin,f_min,flag,sol_exacte,nb_iters)


## Interprétation
 1.Commenter les résultats obtenus, en étudiant notamment les valeurs de $\lambda_k$ et $\mu_k$.
 
 2.Étudier l'influence du paramètre $\tau$ dans la performance de l'algorithme.
 
 3.**Supplémentaire** : 
      Que proposez-vous comme méthode pour la résolution des problèmes avec
      des contraintes à la fois d'égalité et d'inégalité ? Implémenter (si le temps le permet)
      ce nouvel algorithme


## Réponses:
1. As can be seen from the test results,  
   - Properly reducing $\mu_0$ can increase the number of iterations of the algorithm. Too small $\mu_0$ ($\mu_0$=0.1) will cause the algorithm to not iterate and eventually get wrong results. Increasing mu0 will reduce the number of iterations of the algorithm.
   - Increasing or decreasing $\lambda_0$ will not cause drastic changes in the number of iterations.
2. When $\tau$ decreases, the number of iterations will increase, and the accuracy of each iteration will decrease; increasing $\tau$ produces more accurate results